File: frontend/src/types/idms.ts
export interface IdmsSessionStatus {
  authenticated: boolean
  mfa_required: boolean
  message: string
}

export interface IdmsSyncResult {
  report_id: string
  year: number
  rows_inserted: number
  message: string
}

export interface IdmsChargeOff {
  id: string
  report_year: number
  acct_id: string
  borrower: string | null
  date_sold: string | null
  charge_off_date: string | null
  vin: string | null
  year: string | null
  make: string | null
  model: string | null
  original_balance: string
  original_total_balance: string
  total_recovery: string
  current_balance: string
  total_adjusted: string
  repo_method: string | null
  status: string | null
  acct_flags: string | null
  imported_at: string
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export interface IdmsChargeOffMonthly {
  year: number
  month: number
  month_name: string
  count: number
  original_balance: string
  current_balance: string
  total_recovery: string
  total_adjusted: string
}

export interface IdmsDelta {
  value: string
  pct: string
}

export interface IdmsChargeOffOverview {
  year: number
  months_with_data: number[]
  ytd_count: number
  ytd_total_charge_off: string
  ytd_avg_prin_bal: string
  delta_count: IdmsDelta
  delta_total_charge_off: IdmsDelta
  delta_avg_prin_bal: IdmsDelta
  mtd_count: number
  mtd_total_charge_off: string
  mtd_avg_prin_bal: string
  recovery_ratio: string
  gross_co_ratio: string
  annualized_co_ratio: string
  has_portfolio_data: boolean
}

export interface IdmsChargeOffMonthlyDetail {
  year: number
  month: number
  month_name: string
  principal_balance: string | null
  original_balance: string
  count: number
  current_balance: string
  recovery_acv: string
  recovery_ratio: string
  gross_co_ratio: string | null
  months_on_book: string | null
}

export interface IdmsChargeOffKpis {
  year: number
  count: number
  total_original_balance: string
  total_current_balance: string
  total_recovery: string
  total_adjusted: string
  imported_at: string | null
}

File: backend/app/schemas/idms.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Charge Offs
# ---------------------------------------------------------------------------


class IdmsChargeOffBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    report_year: int
    acct_id: str
    borrower: str | None
    date_sold: date | None
    charge_off_date: date | None
    vin: str | None
    year: str | None
    make: str | None
    model: str | None
    original_balance: Decimal
    original_total_balance: Decimal
    total_recovery: Decimal
    current_balance: Decimal
    total_adjusted: Decimal
    repo_method: str | None
    status: str | None
    acct_flags: str | None
    imported_at: datetime


class IdmsChargeOffOut(IdmsChargeOffBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None


class IdmsChargeOffMonthlyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    year: int
    month: int
    month_name: str
    count: int
    original_balance: Decimal
    current_balance: Decimal
    total_recovery: Decimal
    total_adjusted: Decimal


class IdmsChargeOffKpisOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    year: int
    count: int
    total_original_balance: Decimal
    total_current_balance: Decimal
    total_recovery: Decimal
    total_adjusted: Decimal
    imported_at: datetime | None


# ---------------------------------------------------------------------------
# Charge Off Overview (estilo AutoAnalytix)
# ---------------------------------------------------------------------------


class IdmsDeltaOut(BaseModel):
    """Variación contra el mismo rango de meses del año anterior."""

    value: Decimal
    pct: Decimal


class IdmsChargeOffOverviewOut(BaseModel):
    year: int
    months_with_data: list[int]

    ytd_count: int
    ytd_total_charge_off: Decimal
    ytd_avg_prin_bal: Decimal

    delta_count: IdmsDeltaOut
    delta_total_charge_off: IdmsDeltaOut
    delta_avg_prin_bal: IdmsDeltaOut

    mtd_count: int
    mtd_total_charge_off: Decimal
    mtd_avg_prin_bal: Decimal

    recovery_ratio: Decimal
    gross_co_ratio: Decimal
    annualized_co_ratio: Decimal
    # False mientras no haya snapshots de cartera para el rango: sin ellos,
    # Gross C/O y Annualized no se pueden calcular.
    has_portfolio_data: bool


class IdmsChargeOffMonthlyDetailOut(BaseModel):
    year: int
    month: int
    month_name: str
    principal_balance: Decimal | None
    original_balance: Decimal
    count: int
    current_balance: Decimal
    recovery_acv: Decimal
    recovery_ratio: Decimal
    gross_co_ratio: Decimal | None
    months_on_book: Decimal | None


# ---------------------------------------------------------------------------
# Sync / status
# ---------------------------------------------------------------------------


class IdmsSyncIn(BaseModel):
    year: int


class IdmsSyncOut(BaseModel):
    report_id: str
    year: int
    rows_inserted: int
    message: str


class IdmsSessionStatusOut(BaseModel):
    authenticated: bool
    mfa_required: bool = False
    message: str = ""
